torchqdynamics/adaptive.py

```diff
@@ -0,0 +1,182 @@
+
+        See Equation (4.12) and (4.13) of `Hairer et al., Solving Ordinary Differential
+        Equations I (1993), Springer Series in Computational Mathematics` for the
+        detailed steps.
+        """
+        if error == 0:  # no error -> maximally increase the time step
+            return dt * self.max_factor
+
+        # optimal time step
+        dt_opt = dt * error ** (-1.0 / self.order)
+
+        if error <= 1:  # time step accepted -> take next time step at least as large
+            return dt * min(self.max_factor, max(1.0, self.factor * dt_opt))
+
+        if error > 1:  # time step rejected -> reduce next time step
+            return dt * min(0.9, max(self.min_factor, self.factor * dt_opt))
+
+
+class DormandPrince45(AdaptiveSolver):
+    """Dormand-Prince method for adaptive time step ODE integration.
+
+    This is a fifth order solver that uses a fourth order solution to estimate the
+    integration error. It does so using only six function evaluations. See `Dormand and
+    Prince, A family of embedded Runge-Kutta formulae (1980), Journal of Computational
+    and Applied Mathematics`.
+    """
+
+    def __init__(self, *args, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.order = 5
+
+    def build_tableau(self, target):
+        """Build the Butcher tableau of the integrator."""
+        alpha = [1 / 5, 3 / 10, 4 / 5, 8 / 9, 1.0, 1.0, 0.0]
+        beta = [
+            [1 / 5, 0, 0, 0, 0, 0, 0],
+            [3 / 40, 9 / 40, 0, 0, 0, 0, 0],
+            [44 / 45, -56 / 15, 32 / 9, 0, 0, 0, 0],
+            [19372 / 6561, -25360 / 2187, 64448 / 6561, -212 / 729, 0, 0, 0],
+            [9017 / 3168, -355 / 33, 46732 / 5247, 49 / 176, -5103 / 18656, 0, 0],
+            [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0],
+        ]
+        csol = [35 / 384, 0, 500 / 1113, 125 / 192, -2187 / 6784, 11 / 84, 0]
+        cerr = [
+            1951 / 21600,
+            0,
+            22642 / 50085,
+            451 / 720,
+            -12231 / 42400,
+            649 / 6300,
+            1 / 60,
+        ]
+
+        # extract target information
+        dtype = target.dtype
+        real_dtype = to_float(dtype)
+        device = target.device
+
+        # initialize tensors
+        alpha = torch.tensor(alpha, dtype=real_dtype, device=device)
+        beta = torch.tensor(beta, dtype=dtype, device=device)
+        csol = torch.tensor(csol, dtype=dtype, device=device)
+        cerr = torch.tensor(cerr, dtype=dtype, device=device)
+
+        return (alpha, beta, csol, csol - cerr)
+
+    def step(
+        self, f0: Tensor, y0: Tensor, t0: float, dt: float
+    ) -> Tuple[Tensor, Tensor, Tensor]:
+        """Compute a single step of the ODE integration."""
+        # create butcher tableau if not already done
+        if self.tableau is None:
+            self.tableau = self.build_tableau(f0)
+
+        # import butcher tableau
+        alpha, beta, csol, cerr = self.tableau
```

Comment:
It's a bit weird that `csol-cerr` turned into `cerr` here.

Reply:
The point is Dormand-Prince computes 4th order and 5th order solution estimations using the same runge kutta points, and then the difference is an estimate of the error. But the 4th order is only used to compute the error. So anyway, I renamed the above variables for clarity.